### DISCUSSION TITLE
DAI 2024-2025 - Java IOs benchmarking -Jonathan Thiebaud

### DISCUSSION THREAD

**Discussion by JONATHAN-THIEBAUD** [heig-vd-dai-course/heig-vd-dai-course#407] [Show and tell]
https://github.com/heig-vd-dai-course/heig-vd-dai-java-ios.git

# Benchmarking Results

| File Type        | Implementation     | Size    | Execution Time (ms) |
|------------------|--------------------|---------|---------------------|
| binary           | BINARY             | 1B      | 126                 |
| binary           | BINARY             | 1KiB    | 126                 |
| binary           | BINARY             | 1MiB    | 359                 |
| binary           | BINARY             | 5MiB    | 1274                |
| buffered-binary  | BUFFERED_BINARY    | 1B      | 177                 |
| buffered-binary  | BUFFERED_BINARY    | 1KiB    | 130                 |
| buffered-binary  | BUFFERED_BINARY    | 1MiB    | 143                 |
| buffered-binary  | BUFFERED_BINARY    | 5MiB    | 200                 |
| text             | TEXT               | 1B      | 126                 |
| text             | TEXT               | 1KiB    | 121                 |
| text             | TEXT               | 1MiB    | 193                 |
| text             | TEXT               | 5MiB    | 364                 |
| buffered-text    | BUFFERED_TEXT      | 1B      | 122                 |
| buffered-text    | BUFFERED_TEXT      | 1KiB    | 125                 |
| buffered-text    | BUFFERED_TEXT      | 1MiB    | 165                 |
| buffered-text    | BUFFERED_TEXT      | 5MiB    | 189                 |

#### Which type of stream is the most efficient for each use case?
For binary data, **BUFFERED_BINARY** is the most efficient, especially for larger file sizes.
For text data, **BUFFERED_TEXT** performs better as the file size increases.

#### Why is it more efficient than the other types of streams?
Buffered streams are more efficient because they can reduce the number of I/O operations by reading or writing larger chunks of data once a time, instead of performing operations byte by byte.

#### What is the difference between binary data and text data?
Binary data consists of raw bytes that represent information. It is not human-readable but can store more complex data types. Text data, consists of characters and is human-readable. Text files often require encoding, such as UTF-8 or others, to represent characters properly.

#### What is a character encoding?
Character encoding is a system that maps characters to specific binary representations, allowing text to be stored and transmitted as binary data. Common encodings include UTF-8, ASCII, and ISO-8859-1. UTF-8 is widely used because it supports many characters.

#### Why is this benchmark methodology important?
This benchmarking methodology allows us to compare the efficiency of different types of streams in handling various file sizes/types. It helps developers to choose the most efficient implementation for their specific use cases, which is critical for optimizing performance.

**Comment by ludelafo** [comment]
Thank you!

You can always find the solution in the solution repository if you ever need to compare your work with the official one here: https://github.com/heig-vd-dai-course/heig-vd-dai-course-solutions.

Do not hesitate to ask questions if something is not clear, either in this discussion or in the Q&A category of the GitHub Discussions here: https://github.com/orgs/heig-vd-dai-course/discussions/categories/q-a.
